hub/policy/src/main/java/uk/gov/ida/hub/policy/controllogic/AuthnRequestFromTransactionHandler.java

```diff
@@ -122,4 +129,29 @@ public ResponseFromHub getErrorResponseFromHub(SessionId sessionId) {
                 sessionRepository.getStateController(sessionId, ErrorResponsePreparedState.class);
         return stateController.getErrorResponse();
     }
+
+    public AuthnResponseFromCountryContainerDto getAuthnResponseFromCountryContainerDto(SessionId sessionId) {
+        NonMatchingJourneySuccessStateController stateController = (NonMatchingJourneySuccessStateController)
+                sessionRepository.getStateController(sessionId, ResponsePreparedState.class);
+        NonMatchingJourneySuccessState state = stateController.getState();
+        return new AuthnResponseFromCountryContainerDto(
+                state.getCountrySignedResponseContainer().get(),
+                state.getAssertionConsumerServiceUri(),
+                state.getRelayState(),
+                state.getRequestId(),
+                state.getRequestIssuerEntityId(),
+                idGenerator.getId()
+        );
+    }
+
+    public boolean isResponseFromCountryWithUnsignedAssertions(SessionId sessionId) {
+        try {
+            NonMatchingJourneySuccessStateController stateController = (NonMatchingJourneySuccessStateController)
+                    sessionRepository.getStateController(sessionId, ResponsePreparedState.class);
+            NonMatchingJourneySuccessState state = stateController.getState();
+            return state.getCountrySignedResponseContainer().isPresent();
+        } catch (ClassCastException e) {
+            return false;
+        }
```

Comment:
Is this ClassCastException block necessary? We don't catch for it in the method above.

Reply:
We do, because in a success flow there could be two different states at this point. One for a matching flow and one for non matching. If we're in the matching flow, the controller can't be cast to a `NonMatchingJourneySuccessStateController`, so we need this.

Reply:
Or should there be a method `isNonMatchingFlowStateController` or something like that?

Reply:
Have gone for the `instanceof` option. Might be a bit nicer now.